File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/ui/activity/ActListViewModel.java
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.ui.activity;

import androidx.hilt.lifecycle.ViewModelInject;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.model.Activity;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.repository.ActivityRepository;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.State;

public class ActListViewModel extends ViewModel {
    private ActivityRepository activityRepository;
    private MutableLiveData<Activity> selectedActivity = new MutableLiveData<Activity>();
    private MutableLiveData<State<List<Activity>>> allUndoneActivity = new MutableLiveData<State<List<Activity>>>();

    @ViewModelInject
    public ActListViewModel(ActivityRepository activityRepository) {
        this.activityRepository = activityRepository;
        this.allUndoneActivity.postValue(new State<List<Activity>>().init());
        getAllUndoneActivityThread();
    }

    public ActListViewModel(){}

    private void getAllUndoneActivityThread() {
        new Thread(new Runnable() {
            public void run() {
                allUndoneActivity.postValue(new State<List<Activity>>().loading());
                List<Activity> undoneActivity = activityRepository.getAllUndoneActivity();
                allUndoneActivity.postValue(new State<List<Activity>>().success(undoneActivity));
            }
        }).start();
    }

    public LiveData<Activity> getSelectedActivity() {
        return selectedActivity;
    }

    public LiveData<State<List<Activity>>> getAllUndoneActivity() {
        return allUndoneActivity;
    }

    public void selectActivity(Activity selected) {
        selectedActivity.postValue(selected);
    }

    public void checkActivity(Activity selected) {
        new Thread(new Runnable() {
            public void run() {
                activityRepository.updateActivity(selected.getId());
            }
        }).start();
    }
}

File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/ui/activity/ActListViewModel.kt
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.ui.activity

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.model.Activity
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.repository.ActivityRepository
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.State
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ActListViewModel @Inject constructor(private val activityRepository: ActivityRepository) :
    ViewModel() {
    private val selectedActivity = MutableLiveData<Activity>()
    private val allUndoneActivity = MutableLiveData<State<List<Activity>>>()

    fun getSelectedActivity(): LiveData<Activity> {
        return selectedActivity
    }

    fun getAllUndoneActivity(): LiveData<State<List<Activity>>> {
        return allUndoneActivity
    }

    private fun getAllUndoneActivityAsync() {
        viewModelScope.launch(Dispatchers.IO) {
            allUndoneActivity.postValue(State<List<Activity>>().loading())
            val undoneActivity = activityRepository.getAllUndoneActivity()
            allUndoneActivity.postValue(State<List<Activity>>().success(undoneActivity.value))
        }
    }

    fun selectActivity(selected: Activity) {
        selectedActivity.postValue(selected)
    }

    fun checkActivity(selected: Activity) {
        viewModelScope.launch(Dispatchers.IO) {
            activityRepository.updateActivity(selected.id)
        }
    }

    init {
        allUndoneActivity.postValue(State<List<Activity>>().init())
        getAllUndoneActivityAsync()
    }
}